app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/qhybrid/QHybridSupport.java

```diff
@@ -0,0 +1,821 @@
+            case "00002a19-0000-1000-8000-00805f9b34fb": {
+                short level = characteristic.getValue()[0];
+                gbDevice.setBatteryLevel(level);
+
+                gbDevice.setBatteryThresholdPercent((short) 2);
+
+                GBDeviceEventBatteryInfo batteryInfo = new GBDeviceEventBatteryInfo();
+                batteryInfo.level = gbDevice.getBatteryLevel();
+                batteryInfo.state = BatteryState.BATTERY_NORMAL;
+                handleGBDeviceEvent(batteryInfo);
+                break;
+            }
+        }
+
+        return true;
+    }
+
+    private String getModelNameByModelNumber(String modelNumber) {
+        switch (modelNumber) {
+            case "HW.0.0":
+                return "Q Commuter";
+            case "HL.0.0":
+                return "Q Activist";
+        }
+        return "unknwon Q";
+    }
+
+    private void checkFirmwareVersion(String firmwareVersion) {
+        if (!isFirmwareCompatible(firmwareVersion)) {
+            GB.toast("Firmware " + firmwareVersion + " not compatible", Toast.LENGTH_LONG, GB.ERROR);
+            getDevice().setName("incopatible firmware");
+        }
+    }
+
+    private boolean isFirmwareCompatible(String firmwareVersion) {
+        int major = Integer.parseInt(firmwareVersion.substring(6, 7));
+
+        return major == 1;
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
+        switch (characteristic.getUuid().toString()) {
+            case "3dda0004-957f-7d4a-34a6-74696673696d":
+            case "3dda0003-957f-7d4a-34a6-74696673696d": {
+                return handleFileDownloadCharacteristic(characteristic);
+            }
+            case "3dda0007-957f-7d4a-34a6-74696673696d": {
+                return handleFileUploadCharacteristic(characteristic);
+            }
+            case "3dda0002-957f-7d4a-34a6-74696673696d": {
+                return handleBasicCharacteristic(characteristic);
+            }
+            case "3dda0006-957f-7d4a-34a6-74696673696d": {
+                return handleButtonCharacteristic(characteristic);
+            }
+            case "00002a19-0000-1000-8000-00805f9b34fb": {
+                short level = characteristic.getValue()[0];
+                gbDevice.setBatteryLevel(level);
+
+                gbDevice.setBatteryThresholdPercent((short) 2);
+
+                GBDeviceEventBatteryInfo batteryInfo = new GBDeviceEventBatteryInfo();
+                batteryInfo.level = gbDevice.getBatteryLevel();
+                batteryInfo.state = BatteryState.BATTERY_NORMAL;
+                handleGBDeviceEvent(batteryInfo);
+                break;
+            }
+            default: {
+                Log.d("Service", "unknown shit on " + characteristic.getUuid().toString() + ":  " + arrayToString(characteristic.getValue()));
+                try {
+                    File charLog = new File("/sdcard/qFiles/charLog.txt");
```

Comment:
Is this why we need WRITE_SDCARD permission now? Why not use Gadgebridge's data directory?

Reply:
I will address that.